# Phase 2 — Multi-Paradigm Hallucination Detection (AI Agent Conversation Quality Scorer)

**Project:** AI-Agent-Conversation-Quality-Scorer · **Dataset:** HaluEval-QA (Li et al., EMNLP 2023) · **Primary metric:** macro-F1

## The question this session answers
Phase 1 proved HaluEval-QA hides a brutal **length shortcut**: a logistic regression on *answer character count alone*
scores **0.944 raw macro-F1** but collapses to **0.615** once answer lengths are matched between classes. The only
cheap feature that *survived* the length-matched control was **grounding-overlap** (answer→knowledge token overlap):
**0.919 matched macro-F1** — the honest floor.

Phase 2 runs **six paradigms across the full lexical→semantic spectrum** on the *same* frozen split, each scored on
**both** the raw test set and a **rigorous length-matched control**:

| # | Paradigm | Reads the source? | Reads *meaning*? | Trained on HaluEval? |
|---|----------|:--:|:--:|:--:|
| ref | length-only LogReg | ❌ | ❌ | ✅ |
| ref | grounding-overlap threshold (the 0.919 bar) | ✅ (lexical) | ❌ | tuned-threshold only |
| 1 | char-ngram LogReg (answer) | ❌ | ❌ | ✅ |
| 2 | word TF-IDF LogReg (answer ⊕ knowledge) | ✅ (lexical) | ❌ | ✅ |
| 3 | embedding cosine threshold (answer vs knowledge) | ✅ | ✅ (dense) | tuned-threshold only |
| 4 | embedding + semantic features → XGBoost | ✅ | ✅ (dense) | ✅ |
| 5 | **zero-shot NLI cross-encoder** (DeBERTa-v3, answer) | ✅ | ✅ (entailment) | **❌ never** |
| 6 | **zero-shot NLI cross-encoder** (DeBERTa-v3, Q+A) | ✅ | ✅ (entailment) | **❌ never** |

**Concrete bar to beat:** grounding-overlap's **0.919** length-matched macro-F1.
**The headline we are hunting:** *when the length crutch is removed, does a model that reads **meaning** beat one that
counts **lexical overlap** — especially on entity-reuse hallucinations (high overlap, wrong claim) that overlap misses?*

### Research & references that shaped today
1. **Li et al., HaluEval (EMNLP 2023)** — dataset + ChatGPT zero-shot 62.6% acc reference. <https://github.com/RUCAIBox/HaluEval>
2. **Laban et al. / cross-encoder NLI for factual consistency** — NLI entailment between source (premise) and claim
   (hypothesis) is a standard zero-shot grounding check (SummaC, FactCC lineage). Motivates paradigm 5/6.
3. **"The Illusion of Progress" (arXiv:2508.08285, 2025)** — detectors silently exploit length/surface cues; every
   leaderboard here therefore reports raw **and** length-matched.
4. **"Representation-based detectors fail OOD" (arXiv:2509.19372, 2025)** — a model that *never trained on HaluEval*
   (the zero-shot NLI cross-encoder) is the cleanest test of genuine, transferable grounding signal.

In [1]:
import json, os, re, time, warnings
os.environ["TOKENIZERS_PARALLELISM"] = "false"
os.environ["PYTORCH_ENABLE_MPS_FALLBACK"] = "1"   # fall back to CPU for any unsupported MPS op
os.environ["KMP_DUPLICATE_LIB_OK"] = "TRUE"        # torch + xgboost both ship libomp -> allow duplicate (macOS)
os.environ["OMP_NUM_THREADS"] = "1"                # avoid OpenMP thread-init crash (OMP Error #179) in-kernel
warnings.filterwarnings("ignore")
from pathlib import Path
import numpy as np
import pandas as pd
import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt

from sklearn.linear_model import LogisticRegression
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import (f1_score, accuracy_score, balanced_accuracy_score,
                             precision_score, recall_score, roc_auc_score, confusion_matrix)
from scipy.stats import ks_2samp
import sklearn, scipy, torch
print("sklearn", sklearn.__version__, "| torch", torch.__version__,
      "| mps", torch.backends.mps.is_available())

RANDOM_STATE = 42
DEVICE = "mps" if torch.backends.mps.is_available() else "cpu"
print("device:", DEVICE)

def find_root():
    p = Path.cwd()
    for cand in [p, *p.parents]:
        if (cand / "data" / "raw" / "qa_data.json").exists():
            return cand
    return p
ROOT = find_root()
RESULTS = ROOT / "results"; RESULTS.mkdir(exist_ok=True)
CACHE = RESULTS / "phase2_cache"; CACHE.mkdir(exist_ok=True)
plt.rcParams.update({"figure.dpi": 110, "axes.grid": True, "grid.alpha": 0.3})
print("root:", ROOT)

sklearn 1.8.0 | torch 2.11.0 | mps True
device: mps
root: /Users/anthonyrodrigues/Desktop/YC-Portfolio-Projects/AI-Agent-Conversation-Quality-Scorer


## 1. Reload data, features, and the **frozen** Phase-1 split
We reuse the *exact* feature definitions and the frozen `qid` split from Phase 1 so every number is comparable
across phases. `ground_overlap` = fraction of (content) answer tokens that appear in the knowledge passage.

In [2]:
raw_path = ROOT / "data" / "raw" / "qa_data.json"
records = [json.loads(l) for l in raw_path.read_text().splitlines() if l.strip()]
long = []
for qid, r in enumerate(records):
    long.append({"qid": qid, "knowledge": r["knowledge"], "question": r["question"],
                 "answer": r["right_answer"], "label": 0})
    long.append({"qid": qid, "knowledge": r["knowledge"], "question": r["question"],
                 "answer": r["hallucinated_answer"], "label": 1})
df = pd.DataFrame(long).reset_index(drop=True)        # positions 0..19999 are stable -> embedding/NLI cache keys

# ---- features (identical to Phase 1) ----
df["ans_chars"]  = df.answer.str.len()
df["ans_tokens"] = df.answer.str.split().map(len)
df["q_chars"]    = df.question.str.len()
df["k_chars"]    = df.knowledge.str.len()

STOP = set("a an the of to in on at for and or is was were are be been by with as that this it from".split())
def toks(s): return [t for t in re.findall(r"[a-z0-9]+", str(s).lower()) if t not in STOP]
def grounding_overlap(ans, know):
    a = set(toks(ans)); k = set(toks(know))
    return (len(a & k) / len(a)) if a else 0.0
df["ground_overlap"] = [grounding_overlap(a, k) for a, k in zip(df.answer, df.knowledge)]

# ---- frozen split ----
split = json.load(open(RESULTS / "phase1_split_qids.json"))
train_qids, test_qids = set(split["train_qids"]), set(split["test_qids"])
train = df[df.qid.isin(train_qids)].copy()
test  = df[df.qid.isin(test_qids)].copy()
assert not (set(train.qid) & set(test.qid)), "qid leakage!"
LEN_FEATS = ["ans_chars","ans_tokens","q_chars","k_chars"]
print(f"train {len(train)} / {train.qid.nunique()} items   test {len(test)} / {test.qid.nunique()} items")
print("test class balance:", test.label.mean().round(3))
print("mean overlap  grounded:", round(df[df.label==0].ground_overlap.mean(),3),
      " hallucinated:", round(df[df.label==1].ground_overlap.mean(),3))

train 16000 / 8000 items   test 4000 / 2000 items
test class balance: 0.5
mean overlap  grounded: 0.936  hallucinated: 0.612


## 2. A **rigorous** length-matched control (nearest-length pairing)
Phase 1 used a quick per-10-char-bin downsample (KS dropped 0.874 → 0.123). Here we *promote* it: greedily pair each
hallucinated test answer with the **nearest-length grounded** answer (each used once) under a tight character caliper,
so the two classes become near-identical in length. We report KS for the raw test, the Phase-1 bin control, and this
nearest-pair control. Any model that still separates the classes on this set is reading something **other than length**.

In [3]:
import bisect
def nearest_length_match(frame, caliper=8):
    g0 = frame[frame.label==0].sort_values("ans_chars")
    g1 = frame[frame.label==1].sort_values("ans_chars")
    chars0 = g0.ans_chars.tolist(); idx0 = g0.index.tolist()
    keep0, keep1 = [], []
    for c1, i1 in zip(g1.ans_chars.tolist(), g1.index.tolist()):
        if not chars0: break
        p = bisect.bisect_left(chars0, c1)
        best = None
        for j in (p-1, p, p+1):
            if 0 <= j < len(chars0):
                d = abs(chars0[j] - c1)
                if best is None or d < best[0]: best = (d, j)
        d, j = best
        if d <= caliper:
            keep1.append(i1); keep0.append(idx0[j])
            del chars0[j]; del idx0[j]
    return frame.loc[keep0 + keep1]

def bin_match(frame, width=10, seed=RANDOM_STATE):           # Phase-1 method, for comparison
    rng = np.random.default_rng(seed); keep = []
    edges = np.arange(0, int(frame.ans_chars.max())+width+1, width)
    lbin = np.digitize(frame.ans_chars, edges)
    for b in np.unique(lbin):
        sub = frame[lbin == b]
        a, c = sub.index[sub.label==0], sub.index[sub.label==1]
        k = min(len(a), len(c))
        if k: keep += list(rng.choice(a,k,False)) + list(rng.choice(c,k,False))
    return frame.loc[keep]

matched     = nearest_length_match(test, caliper=8)
matched_bin = bin_match(test, 10)
def ks(fr): return ks_2samp(fr[fr.label==0].ans_chars, fr[fr.label==1].ans_chars).statistic
print(f"raw test            n={len(test):5d}  KS(ans_chars)={ks(test):.3f}")
print(f"bin control (P1)    n={len(matched_bin):5d}  KS={ks(matched_bin):.3f}")
print(f"nearest-pair (P2)   n={len(matched):5d}  KS={ks(matched):.3f}  "
      f"({int((matched.label==0).sum())} grounded / {int((matched.label==1).sum())} hallucinated)")
print("mean ans_chars  grounded:", round(matched[matched.label==0].ans_chars.mean(),1),
      " hallucinated:", round(matched[matched.label==1].ans_chars.mean(),1))

raw test            n= 4000  KS(ans_chars)=0.874
bin control (P1)    n=  522  KS=0.123
nearest-pair (P2)   n=  572  KS=0.122  (286 grounded / 286 hallucinated)
mean ans_chars  grounded: 30.2  hallucinated: 29.4


## 3. Evaluation harness + Phase-1 reference models
`SCORERS[name](frame) -> (y_pred, y_score)` lets every model be re-scored on any frame (raw test or the matched
control) with one call. We carry forward two anchors: the **length-only** shortcut (should collapse on matched) and the
**grounding-overlap** threshold (the **0.919** bar to beat).

In [4]:
def evaluate(name, fr, y_pred, y_score, split):
    y = fr.label
    return {"model": name, "split": split, "n": int(len(fr)),
            "accuracy": round(accuracy_score(y, y_pred), 4),
            "macro_f1": round(f1_score(y, y_pred, average="macro"), 4),
            "balanced_acc": round(balanced_accuracy_score(y, y_pred), 4),
            "precision_hallu": round(precision_score(y, y_pred, pos_label=1, zero_division=0), 4),
            "recall_hallu": round(recall_score(y, y_pred, pos_label=1, zero_division=0), 4),
            "roc_auc": round(roc_auc_score(y, y_score), 4) if y_score is not None and len(set(y))>1 else None}

SCORERS = {}      # name -> fn(frame) -> (pred, score)
ORDER = []        # display order

def register(name, fn):
    SCORERS[name] = fn
    if name not in ORDER: ORDER.append(name)

def tune_threshold(score_train, y_train, predict_high_is_hallu=True):
    """return threshold maximizing train macro-F1; pred = (score >= t) if high_is_hallu."""
    best_t, best_f1 = 0.5, -1
    for t in np.linspace(score_train.min(), score_train.max(), 201):
        pred = (score_train >= t).astype(int) if predict_high_is_hallu else (score_train < t).astype(int)
        f = f1_score(y_train, pred, average="macro")
        if f > best_f1: best_f1, best_t = f, t
    return best_t, best_f1

# --- length-only LogReg (the shortcut) ---
len_pipe = Pipeline([("sc", StandardScaler()),
                     ("lr", LogisticRegression(max_iter=2000, random_state=RANDOM_STATE))]).fit(train[LEN_FEATS], train.label)
def score_len(fr):
    s = len_pipe.predict_proba(fr[LEN_FEATS])[:,1]; return (s>=.5).astype(int), s
register("length_only_logreg", score_len)

# --- grounding-overlap threshold (the 0.919 bar): hallucinated when overlap is LOW ---
ov_t, ov_f1 = tune_threshold(train.ground_overlap.values, train.label.values, predict_high_is_hallu=False)
def score_overlap(fr):
    s = 1 - fr.ground_overlap.values; return (fr.ground_overlap.values < ov_t).astype(int), s
register("grounding_overlap_threshold", score_overlap)
print(f"overlap threshold tuned on train: overlap < {ov_t:.3f}  (train macroF1={ov_f1:.4f})")

overlap threshold tuned on train: overlap < 0.960  (train macroF1=0.9256)


## 4. Paradigm 1 — char-ngram LogReg on the answer (pure surface style)
Character 3–5 grams capture *how an answer is written* (fluent ChatGPT sentence vs. terse entity span) with zero
access to the source. Expectation: strong on raw, partial collapse on matched (style correlates with length/fluency).

In [5]:
char_pipe = Pipeline([
    ("tf", TfidfVectorizer(analyzer="char_wb", ngram_range=(3,5), min_df=3, max_features=60000)),
    ("lr", LogisticRegression(max_iter=3000, C=1.0, random_state=RANDOM_STATE))]).fit(train.answer, train.label)
def score_char(fr):
    s = char_pipe.predict_proba(fr.answer)[:,1]; return (s>=.5).astype(int), s
register("char_ngram_logreg", score_char)
print("char-ngram vocab:", len(char_pipe.named_steps["tf"].vocabulary_))
_p,_s = score_char(test); print("char raw test macroF1:", round(f1_score(test.label,_p,average='macro'),4))

char-ngram vocab: 42591
char raw test macroF1: 0.9515


## 5. Paradigm 2 — word TF-IDF LogReg on answer ⊕ knowledge (lexical, *reads the source*)
Phase 1 found that adding the *question* HURT (shared across a pair). Here we instead concatenate the **knowledge**,
giving the linear model the lexical material to check the answer against — a learned cousin of grounding-overlap.

In [6]:
def ak(fr): return (fr.answer + " [SEP] " + fr.knowledge).tolist()
word_pipe = Pipeline([
    ("tf", TfidfVectorizer(ngram_range=(1,2), min_df=3, max_features=80000)),
    ("lr", LogisticRegression(max_iter=3000, C=1.0, random_state=RANDOM_STATE))]).fit(ak(train), train.label)
def score_word(fr):
    s = word_pipe.predict_proba(ak(fr))[:,1]; return (s>=.5).astype(int), s
register("word_tfidf_ans_know", score_word)
_p,_s = score_word(test); print("word A+K raw test macroF1:", round(f1_score(test.label,_p,average='macro'),4))

word A+K raw test macroF1: 0.6951


## 6. Embedding mini-leaderboard (which encoder gives the best grounding signal?)
We embed each **answer** and its **knowledge** with three off-the-shelf sentence encoders, take the cosine similarity
as a single grounding score, tune a threshold on train, and rank by **matched** macro-F1. Embeddings are cached to
`results/phase2_cache/` so reruns are instant. (e5 uses its `query:`/`passage:` prefixes; MiniLM & BGE encode raw.)

In [7]:
from sentence_transformers import SentenceTransformer

EMB_MODELS = {
    "MiniLM-L6":   "sentence-transformers/all-MiniLM-L6-v2",
    "bge-small":   "BAAI/bge-small-en-v1.5",
    "e5-base":     "intfloat/e5-base-v2",
}
def enc_cached(tag, model_id, texts, prefix=""):
    fp = CACHE / f"emb_{tag}.npy"
    if fp.exists():
        arr = np.load(fp)
        if arr.shape[0] == len(texts): return arr
    m = SentenceTransformer(model_id, device=DEVICE)
    arr = m.encode([prefix + t for t in texts], normalize_embeddings=True,
                   batch_size=128, show_progress_bar=False).astype(np.float32)
    np.save(fp, arr); return arr

answers   = df.answer.tolist()
knowledge = df.knowledge.tolist()
emb_lb = []
COS = {}                                   # tag -> per-row cosine (len(df))
for tag, mid in EMB_MODELS.items():
    qp = "query: " if tag=="e5-base" else ""
    pp = "passage: " if tag=="e5-base" else ""
    t0 = time.time()
    ea = enc_cached(f"{tag}_ans", mid, answers, qp)
    ek = enc_cached(f"{tag}_know", mid, knowledge, pp)
    cos = np.sum(ea * ek, axis=1)          # both normalized -> dot = cosine
    COS[tag] = cos
    tr_cos, te_cos = cos[train.index.values], cos[test.index.values]
    t, f1tr = tune_threshold(tr_cos, train.label.values, predict_high_is_hallu=False)  # low cos -> hallu
    # eval on raw test + matched
    def mk_pred(c): return (c < t).astype(int)
    rawf1 = f1_score(test.label, mk_pred(te_cos), average="macro")
    mf1   = f1_score(matched.label, mk_pred(cos[matched.index.values]), average="macro")
    emb_lb.append({"encoder": tag, "dim": ea.shape[1], "thr": round(float(t),3),
                   "raw_macroF1": round(rawf1,4), "matched_macroF1": round(mf1,4),
                   "encode_s": round(time.time()-t0,1)})
emb_lb = pd.DataFrame(emb_lb).sort_values("matched_macroF1", ascending=False).reset_index(drop=True)
WINNER = emb_lb.iloc[0].encoder
display(emb_lb)
print("winning encoder (by matched macro-F1):", WINNER)

,encoder,dim,thr,raw_macroF1,matched_macroF1,encode_s
0,e5-base,768,0.681,0.3359,0.3411,0.4
1,MiniLM-L6,384,0.966,0.3333,0.3333,0.4
2,bge-small,384,0.332,0.3332,0.3333,0.4


winning encoder (by matched macro-F1): e5-base


## 7. Paradigm 3 — embedding cosine threshold (winner) + max-sentence similarity
The pure dense-semantic analogue of grounding-overlap: predict *hallucination* when answer↔knowledge cosine is low.
We also compute **max sentence-level similarity** (best-supporting-sentence) with the winning encoder, since a short
answer can be entailed by a single sentence even if its similarity to the *whole* passage is diluted.

In [8]:
winner_id = EMB_MODELS[WINNER]
qp = "query: " if WINNER=="e5-base" else ""
pp = "passage: " if WINNER=="e5-base" else ""
df["emb_cos"] = COS[WINNER]

# max sentence-level similarity (winner encoder), cached
def split_sents(txt):
    s = re.split(r"(?<=[.!?])\s+", str(txt).strip())
    return [x for x in s if x.strip()] or [str(txt)]

ms_fp = CACHE / f"maxsent_{WINNER}.npy"
if ms_fp.exists():
    df["emb_maxsent"] = np.load(ms_fp)
else:
    m = SentenceTransformer(winner_id, device=DEVICE)
    uniq = {q: split_sents(k) for q, k in zip(df.qid, df.knowledge)}
    flat, spans, cur = [], {}, 0
    for q, sents in uniq.items():
        spans[q] = (cur, cur+len(sents)); cur += len(sents); flat += [pp+s for s in sents]
    sent_emb = m.encode(flat, normalize_embeddings=True, batch_size=128, show_progress_bar=False).astype(np.float32)
    ans_emb  = enc_cached(f"{WINNER}_ans", winner_id, df.answer.tolist(), qp)
    maxsent = np.empty(len(df), np.float32)
    for pos, (q, ae) in enumerate(zip(df.qid, ans_emb)):
        a, b = spans[q]; maxsent[pos] = float(np.max(sent_emb[a:b] @ ae))
    df["emb_maxsent"] = maxsent; np.save(ms_fp, maxsent)

# cosine-threshold scorer (winner) — pull columns from df by row index (robust to stale train/test copies)
ct, ctf1 = tune_threshold(df.loc[train.index,"emb_cos"].values, train.label.values, predict_high_is_hallu=False)
def score_cos(fr):
    c = df.loc[fr.index, "emb_cos"].values
    return (c < ct).astype(int), 1 - c
register(f"embed_cosine_{WINNER}", score_cos)
print(f"cosine threshold ({WINNER}): cos < {ct:.3f}  train macroF1={ctf1:.4f}")
print("mean emb_maxsent  grounded:", round(df[df.label==0].emb_maxsent.mean(),3),
      " hallucinated:", round(df[df.label==1].emb_maxsent.mean(),3))

cosine threshold (e5-base): cos < 0.681  train macroF1=0.3352
mean emb_maxsent  grounded: 0.806  hallucinated: 0.844


## 8. Paradigm 4 — XGBoost on **semantic features only** (no length)
A learned head over three *content* signals — `emb_cos`, `emb_maxsent`, `ground_overlap` — deliberately **excluding**
length. If this holds up on the matched control, it confirms the signal is grounding, not surface. We also fit a
second variant that *adds* length features, to foreshadow the Phase-5 ablation (expect: big raw gain, matched collapse).

In [9]:
from xgboost import XGBClassifier
SEM = ["emb_cos","emb_maxsent","ground_overlap"]
def fit_xgb(feats):
    clf = XGBClassifier(n_estimators=400, max_depth=4, learning_rate=0.05,
                        subsample=0.9, colsample_bytree=0.9, eval_metric="logloss",
                        random_state=RANDOM_STATE, n_jobs=1)     # n_jobs=1: avoid OpenMP thread-init crash
    return clf.fit(df.loc[train.index, feats], train.label)      # pull features from df by row index

xgb_sem = fit_xgb(SEM)
def score_xgb_sem(fr):
    s = xgb_sem.predict_proba(df.loc[fr.index, SEM])[:,1]; return (s>=.5).astype(int), s
register("embed_xgb_semantic", score_xgb_sem)

xgb_len = fit_xgb(SEM + LEN_FEATS)
def score_xgb_len(fr):
    s = xgb_len.predict_proba(df.loc[fr.index, SEM+LEN_FEATS])[:,1]; return (s>=.5).astype(int), s
register("embed_xgb_plus_length", score_xgb_len)

imp = dict(zip(SEM, np.round(xgb_sem.feature_importances_,3)))
print("semantic-only XGB feature importance:", imp)
_p,_s = score_xgb_sem(test); print("xgb-semantic raw test macroF1:", round(f1_score(test.label,_p,average='macro'),4))

semantic-only XGB feature importance: {'emb_cos': np.float32(0.041), 'emb_maxsent': np.float32(0.017), 'ground_overlap': np.float32(0.942)}
xgb-semantic raw test macroF1: 0.9582


## 9. Paradigms 5 & 6 — **zero-shot NLI cross-encoder** (DeBERTa-v3, never trained on HaluEval)
The cleanest meaning test. Premise = **knowledge**, hypothesis = **answer** (paradigm 5) or **question + answer**
(paradigm 6, which turns a bare entity span into a checkable proposition). The model outputs P(contradiction),
P(entailment), P(neutral); our **hallucination score = 1 − P(entailment)**. We tune only a decision threshold on a
stratified train-dev slice (the model weights are frozen). NLI scores are cached so this runs once.

In [10]:
from sentence_transformers import CrossEncoder
NLI_ID = "cross-encoder/nli-deberta-v3-base"

# rows we need NLI for: all test + a stratified train-dev slice (only to tune a decision threshold)
rng = np.random.default_rng(RANDOM_STATE)
dev_pos = np.concatenate([
    rng.choice(train.index[train.label==0].values, 500, replace=False),
    rng.choice(train.index[train.label==1].values, 500, replace=False)])
need_pos = np.unique(np.concatenate([test.index.values, dev_pos]))

def nli_scores(variant, chunk=1024, batch_size=64):
    """variant 'ans'|'qa'. (N,3) softmax [contra, entail, neutral]. Chunked + incrementally cached -> resumable."""
    fp = CACHE / f"nli_{variant}.npy"
    full = np.load(fp) if fp.exists() else np.full((len(df),3), np.nan, np.float32)
    todo = [int(p) for p in need_pos if np.isnan(full[p,0])]
    if not todo:
        print(f"NLI[{variant}] fully cached ({len(need_pos)} rows)"); return full
    ce = CrossEncoder(NLI_ID, device=DEVICE, max_length=288)
    t0 = time.time(); done = 0
    for s in range(0, len(todo), chunk):
        idx = todo[s:s+chunk]
        if variant == "ans":
            pairs = [(df.knowledge.iat[p], df.answer.iat[p]) for p in idx]
        else:
            pairs = [(df.knowledge.iat[p], df.question.iat[p] + " " + df.answer.iat[p]) for p in idx]
        sc = ce.predict(pairs, apply_softmax=True, batch_size=batch_size, show_progress_bar=False)
        full[idx] = np.asarray(sc, np.float32); np.save(fp, full)         # save each chunk -> resumable
        done += len(idx); print(f"  NLI[{variant}] {done}/{len(todo)}  {time.time()-t0:.0f}s", flush=True)
    return full

ENTAIL = 1   # id2label = {0:contradiction, 1:entailment, 2:neutral}
nli_ans = nli_scores("ans")
nli_qa  = nli_scores("qa")
df["nli_ans_halu"] = 1 - nli_ans[:,ENTAIL]    # high = likely hallucinated
df["nli_qa_halu"]  = 1 - nli_qa[:,ENTAIL]
g = df.loc[test.index]
print("mean 1-P(entail) [ans]  grounded:", round(np.nanmean(g[g.label==0].nli_ans_halu),3),
      " hallucinated:", round(np.nanmean(g[g.label==1].nli_ans_halu),3))

NLI[ans] fully cached (5000 rows)
NLI[qa] fully cached (5000 rows)
mean 1-P(entail) [ans]  grounded: 0.792  hallucinated: 0.742


### Calibrate NLI thresholds on the dev slice and register both variants
We also log the **zero-threshold** (pure argmax: entailment vs. not) accuracy as a no-calibration reference.

In [11]:
dev = df.loc[dev_pos]
for variant, col in [("ans","nli_ans_halu"), ("qa","nli_qa_halu")]:
    t, f1tr = tune_threshold(dev[col].values, dev.label.values, predict_high_is_hallu=True)
    argmax_pred = (dev[col].values >= 0.5).astype(int)   # >=0.5 means P(entail)<0.5 -> not-entailed -> hallu
    print(f"NLI[{variant}] tuned thr={t:.3f} (dev macroF1={f1tr:.4f}) | "
          f"argmax(@0.5) dev macroF1={f1_score(dev.label, argmax_pred, average='macro'):.4f}")
    def make(col=col, t=t):
        def fn(fr):
            s = df.loc[fr.index, col].values; return (s >= t).astype(int), s   # pull from df by row index
        return fn
    register(f"nli_zeroshot_{variant}", make())

NLI[ans] tuned thr=0.995 (dev macroF1=0.5153) | argmax(@0.5) dev macroF1=0.4385


NLI[qa] tuned thr=0.990 (dev macroF1=0.6742) | argmax(@0.5) dev macroF1=0.6037


## 10. Dual leaderboard — raw vs. length-matched (ranked by **matched** macro-F1)
Every registered paradigm scored on the same raw test set and the same nearest-pair matched control.

In [12]:
rows_raw, rows_mat = [], []
for name in ORDER:
    fn = SCORERS[name]
    pr, sr = fn(test);    rows_raw.append(evaluate(name, test, pr, sr, "raw"))
    pm, sm = fn(matched); rows_mat.append(evaluate(name, matched, pm, sm, "length_matched"))
raw_b = pd.DataFrame(rows_raw); mat_b = pd.DataFrame(rows_mat)
comp = (raw_b[["model","macro_f1","recall_hallu","roc_auc"]]
        .merge(mat_b[["model","macro_f1","recall_hallu"]], on="model", suffixes=("_raw","_matched")))
comp["drop"] = (comp.macro_f1_raw - comp.macro_f1_matched).round(4)
comp = comp.sort_values("macro_f1_matched", ascending=False).reset_index(drop=True)
comp.insert(0, "rank", comp.index + 1)
pd.set_option("display.width", 160)
display(comp.rename(columns={"macro_f1_raw":"raw_F1","macro_f1_matched":"matched_F1",
                             "recall_hallu_raw":"raw_recall_H","recall_hallu_matched":"mat_recall_H",
                             "roc_auc":"raw_AUC"}))
BAR = 0.919
best = comp.iloc[0]
print(f"\nBAR (Phase-1 grounding-overlap) matched macro-F1 = {BAR}")
print(f"Phase-2 best on matched control: {best.model} = {best.macro_f1_matched:.4f} "
      f"({'BEATS' if best.macro_f1_matched>BAR else 'does NOT beat'} the bar)")

,rank,model,raw_F1,raw_recall_H,raw_AUC,matched_F1,mat_recall_H,drop
0,1,grounding_overlap_threshold,0.9252,0.9125,0.8989,0.9244,0.8531,0.0008
1,2,embed_xgb_semantic,0.9582,0.9265,0.9803,0.8890,0.8007,0.0692
2,3,embed_xgb_plus_length,0.9742,0.9705,0.9974,0.8671,0.8741,0.1071
3,4,char_ngram_logreg,0.9515,0.9320,0.9793,0.7636,0.6713,0.1879
4,5,nli_zeroshot_qa,0.6427,0.5890,0.6711,0.6866,0.6469,-0.0439
5,6,word_tfidf_ans_know,0.6951,0.7145,0.7683,0.6311,0.6189,0.0640
6,7,length_only_logreg,0.9437,0.9345,0.9713,0.6174,0.5559,0.3263
7,8,nli_zeroshot_ans,0.5197,0.6060,0.5406,0.5885,0.6294,-0.0688
8,9,embed_cosine_e5-base,0.3359,0.0030,0.2291,0.3411,0.0070,-0.0052



BAR (Phase-1 grounding-overlap) matched macro-F1 = 0.919
Phase-2 best on matched control: grounding_overlap_threshold = 0.9244 (BEATS the bar)


## 11. The real question — **entity-reuse hallucinations** (high overlap, wrong claim)
Grounding-overlap's blind spot is the hallucination that *reuses the passage's words* but asserts something false
(high `ground_overlap`, label=1). These are exactly where a model must read **meaning**, not lexical overlap. We
isolate the test hallucinations with above-median overlap and measure each paradigm's **recall** on them.

In [13]:
halu = test[test.label==1]
ov_med = halu.ground_overlap.median()
hard = halu[halu.ground_overlap >= ov_med]    # entity-reuse hallucinations
easy = halu[halu.ground_overlap <  ov_med]
print(f"entity-reuse (high-overlap) hallucinations: n={len(hard)} (overlap>={ov_med:.2f}); "
      f"low-overlap: n={len(easy)}")

reuse_rows = []
for name in ORDER:
    fn = SCORERS[name]
    rec_hard = recall_score(hard.label, fn(hard)[0], pos_label=1, zero_division=0)
    rec_easy = recall_score(easy.label, fn(easy)[0], pos_label=1, zero_division=0)
    reuse_rows.append({"model": name, "recall_easy_halu": round(rec_easy,3),
                       "recall_entity_reuse_halu": round(rec_hard,3),
                       "gap": round(rec_easy-rec_hard,3)})
reuse = pd.DataFrame(reuse_rows).sort_values("recall_entity_reuse_halu", ascending=False).reset_index(drop=True)
display(reuse)
top = reuse.iloc[0]
print(f"\nBest at catching entity-reuse hallucinations: {top.model} "
      f"(recall={top.recall_entity_reuse_halu:.3f})")

entity-reuse (high-overlap) hallucinations: n=1054 (overlap>=0.62); low-overlap: n=946


,model,recall_easy_halu,recall_entity_reuse_halu,gap
0,length_only_logreg,0.895,0.970,-0.074
1,char_ngram_logreg,0.897,0.963,-0.066
2,embed_xgb_plus_length,0.997,0.947,0.050
3,embed_xgb_semantic,0.974,0.884,0.089
4,grounding_overlap_threshold,1.000,0.834,0.166
5,word_tfidf_ans_know,0.708,0.720,-0.012
6,nli_zeroshot_ans,0.740,0.486,0.254
7,nli_zeroshot_qa,0.714,0.477,0.236
8,embed_cosine_e5-base,0.006,0.000,0.006



Best at catching entity-reuse hallucinations: length_only_logreg (recall=0.970)


### Where overlap is *fooled* but meaning saves the day
Among the entity-reuse hallucinations that **grounding-overlap misclassifies as grounded**, what fraction does the
best NLI paradigm catch? This is the concrete "meaning > lexical" payoff.

In [14]:
ov_pred_hard = SCORERS["grounding_overlap_threshold"](hard)[0]
fooled = hard[ov_pred_hard == 0]                      # overlap said "grounded", truth = hallucinated
print(f"entity-reuse hallucinations overlap got WRONG: {len(fooled)} / {len(hard)}")
rescue = {}
for name in ORDER:
    if name == "grounding_overlap_threshold": continue
    caught = recall_score(fooled.label, SCORERS[name](fooled)[0], pos_label=1, zero_division=0)
    rescue[name] = round(caught, 3)
rescue = pd.Series(rescue).sort_values(ascending=False)
print("\nfraction of overlap's entity-reuse misses RESCUED by each paradigm:")
print(rescue.to_string())

entity-reuse hallucinations overlap got WRONG: 175 / 1054

fraction of overlap's entity-reuse misses RESCUED by each paradigm:
length_only_logreg       0.897
char_ngram_logreg        0.886
embed_xgb_plus_length    0.680
word_tfidf_ans_know      0.566
embed_xgb_semantic       0.303
nli_zeroshot_qa          0.206
nli_zeroshot_ans         0.200
embed_cosine_e5-base     0.000


## 12. Figures

In [15]:
# Fig 1 — dual leaderboard (comp uses macro_f1_raw / macro_f1_matched)
fig, ax = plt.subplots(figsize=(10,5.5))
o = comp.sort_values("macro_f1_matched")
y = np.arange(len(o)); h = 0.4
ax.barh(y+h/2, o.macro_f1_raw, h, label="raw test", color="#9ecae1")
ax.barh(y-h/2, o.macro_f1_matched, h, label="length-matched", color="#3182bd")
ax.axvline(0.919, ls="--", c="crimson", lw=1.5, label="Phase-1 bar 0.919 (matched)")
ax.set_yticks(y); ax.set_yticklabels(o.model, fontsize=9)
ax.set_xlabel("macro-F1"); ax.set_xlim(0.3,1.0)
ax.set_title("Phase 2 — raw vs. length-matched macro-F1 (the matched bar is the truth)")
ax.legend(loc="lower right", fontsize=8)
plt.tight_layout(); plt.savefig(RESULTS/"phase2_dual_leaderboard.png", dpi=130); plt.show(); plt.close()

# Fig 2 — entity-reuse recall
fig, ax = plt.subplots(figsize=(10,5.5))
o2 = reuse.sort_values("recall_entity_reuse_halu")
y = np.arange(len(o2)); h=0.4
ax.barh(y+h/2, o2.recall_easy_halu, h, label="low-overlap hallu", color="#a1d99b")
ax.barh(y-h/2, o2.recall_entity_reuse_halu, h, label="entity-reuse hallu (high overlap)", color="#31a354")
ax.set_yticks(y); ax.set_yticklabels(o2.model, fontsize=9)
ax.set_xlabel("recall on hallucinations"); ax.set_xlim(0,1.02)
ax.set_title("Who catches entity-reuse hallucinations (high overlap, wrong claim)?")
ax.legend(loc="lower left", fontsize=8)
plt.tight_layout(); plt.savefig(RESULTS/"phase2_entity_reuse_recall.png", dpi=130); plt.show(); plt.close()

# Fig 3 — NLI separation
fig, ax = plt.subplots(1,2, figsize=(11,4))
for a, col, ttl in [(ax[0],"nli_ans_halu","NLI hypothesis = answer"),
                    (ax[1],"nli_qa_halu","NLI hypothesis = question + answer")]:
    g = df.loc[test.index]
    a.hist(g[g.label==0][col], bins=30, alpha=0.6, label="grounded", color="#3182bd", density=True)
    a.hist(g[g.label==1][col], bins=30, alpha=0.6, label="hallucinated", color="#e6550d", density=True)
    a.set_title(ttl); a.set_xlabel("1 - P(entailment)"); a.legend(fontsize=8)
plt.tight_layout(); plt.savefig(RESULTS/"phase2_nli_separation.png", dpi=130); plt.show(); plt.close()
print("saved 3 figures to results/")

saved 3 figures to results/


## 13. Persist Phase-2 results

In [16]:
comp.to_csv(RESULTS/"phase2_model_comparison.csv", index=False)
reuse.to_csv(RESULTS/"phase2_entity_reuse.csv", index=False)
emb_lb.to_csv(RESULTS/"phase2_embedding_leaderboard.csv", index=False)

mp = RESULTS/"metrics.json"
existing = json.load(open(mp)) if mp.exists() else {}
existing["phase2"] = {
    "dataset": "HaluEval-QA",
    "primary_metric": "macro_f1",
    "bar_matched_macro_f1": 0.919,
    "matched_control": {"method": "nearest-length pairing (caliper 8 chars)",
                        "n": int(len(matched)), "ks": round(float(ks(matched)),3),
                        "ks_raw": round(float(ks(test)),3), "ks_bin_p1": round(float(ks(matched_bin)),3)},
    "embedding_leaderboard": emb_lb.to_dict("records"),
    "winner_encoder": WINNER,
    "leaderboard_raw": rows_raw,
    "leaderboard_matched": rows_mat,
    "comparison": comp.to_dict("records"),
    "entity_reuse_recall": reuse.to_dict("records"),
    "best_matched_model": comp.iloc[0].model,
    "best_matched_macro_f1": float(comp.iloc[0].macro_f1_matched),
}
json.dump(existing, open(mp,"w"), indent=2)
print("saved -> results/metrics.json keys:", list(existing.keys()))
print("Phase-2 matched leaderboard (top 5):")
print(comp[["rank","model","macro_f1_matched","macro_f1_raw","drop"]].head(5).to_string(index=False))

saved -> results/metrics.json keys: ['phase1', 'phase2']
Phase-2 matched leaderboard (top 5):
 rank                       model  macro_f1_matched  macro_f1_raw    drop
    1 grounding_overlap_threshold            0.9244        0.9252  0.0008
    2          embed_xgb_semantic            0.8890        0.9582  0.0692
    3       embed_xgb_plus_length            0.8671        0.9742  0.1071
    4           char_ngram_logreg            0.7636        0.9515  0.1879
    5             nli_zeroshot_qa            0.6866        0.6427 -0.0439


## 14. Key findings

**Dual leaderboard, ranked by length-matched macro-F1 (the honest metric):**

| Rank | Paradigm | raw F1 | **matched F1** | raw→matched drop |
|---|---|---:|---:|---:|
| 1 | **grounding_overlap_threshold** (1-line lexical rule) | 0.9252 | **0.9244** | **0.0008** |
| 2 | embed_xgb_semantic (overlap + 2 emb feats) | 0.9582 | 0.8890 | 0.069 |
| 3 | embed_xgb_plus_length | 0.9742 | 0.8671 | 0.107 |
| 4 | char_ngram_logreg (surface style) | 0.9515 | 0.7636 | 0.188 |
| 5 | nli_zeroshot_qa (DeBERTa-v3) | 0.6427 | 0.6866 | −0.044 |
| 6 | word_tfidf_ans⊕know | 0.6951 | 0.6311 | 0.064 |
| 7 | length_only_logreg (the shortcut) | 0.9437 | 0.6174 | 0.326 |
| 8 | nli_zeroshot_ans (DeBERTa-v3) | 0.5197 | 0.5885 | −0.069 |
| 9 | embed_cosine_e5-base (dense similarity) | 0.3359 | 0.3411 | −0.005 |

1. **The dumb 1-line lexical-overlap rule is still undefeated — and it's almost perfectly length-invariant.**
   grounding-overlap tops the matched leaderboard at **0.9244** (beats the 0.919 bar) with a raw→matched drop of
   **0.0008**. Six richer paradigms — three sentence-embedding models and a 184M-param zero-shot NLI cross-encoder —
   *all* fail to beat it.

2. **THE SURPRISE — my hypothesis was wrong: zero-shot NLI is among the *worst*, not the best.** I expected an
   entailment model (premise=knowledge, hypothesis=answer) to *read meaning* and beat lexical overlap, especially on
   entity-reuse hallucinations. Instead NLI-QA = 0.687, NLI-ans = 0.589 matched — far below 0.924. The mechanism is
   damning: **mean P(entailment) is *higher* for hallucinations (0.258) than for grounded answers (0.208)** — the
   signal is *inverted*. HaluEval's grounded answers are bare entity spans ("Arthur's Magazine") that NLI reads as
   *non-entailed* (a fragment isn't a proposition), while the fluent hallucinations reuse the passage's entities and
   read as *entailed*. NLI entailment is confounded by **answer form**, not grounding. Reframing the hypothesis as
   question+answer (paradigm 6) partially repairs it (dev macro-F1 0.515 → 0.674), but it's still nowhere near overlap.

3. **Dense embedding cosine is essentially useless here (matched F1 ≈ 0.34 ≈ the 0.333 floor; recall 0.007).** All
   three encoders sit at chance. Topical similarity ≠ factual grounding: both the right answer and the hallucination
   are *about* the same passage, so cosine-to-passage cannot separate them. e5-base "won" the encoder mini-leaderboard
   only by a rounding artifact above the floor.

4. **XGBoost agrees overlap is THE signal.** Semantic-only feature importance: `ground_overlap` **0.942**, `emb_cos`
   0.041, `emb_maxsent` 0.017 — the tree discards the embeddings. It edges higher on raw (0.958) by leaning on the
   embeddings a little, but that edge does not survive length matching (0.889 < overlap's 0.924).

5. **No free lunch on the hard cases.** Of the 1,054 entity-reuse (high-overlap) hallucinations, overlap misclassifies
   175 (near-verbatim reuse, overlap ≥ 0.96). Those are rescued mostly by **length** (0.90) and **char-ngram** (0.89) —
   i.e. they're long fluent rewrites caught by *surface* cues that the matched control invalidates — and barely by the
   meaning models (NLI 0.20, embeddings 0.00). On HaluEval-QA the cases lexical overlap misses are recoverable only via
   the very shortcut we're trying to remove. **A real meaning model for this task must be trained on the claim↔passage
   relation, not used zero-shot — the target for Phases 3–5.**

6. **Rigorous matched control.** Nearest-length pairing (caliper 8): KS 0.874 → **0.122**, n=572 — comparable KS to
   Phase-1's 10-char-bin (0.123) but deterministic and slightly larger.